src/lib/chart/dendrogram.typ

```diff
@@ -0,0 +1,312 @@
+  if (x-ticks == auto) {
+    // Pre-calculate order of leaf indices
+    let x-counter = 0
+    let ticks = ()
+
+    for (idx, entry) in data.enumerate() {
+      let x1 = entry.at(x1-key)
+      let x2 = entry.at(x2-key)
+
+      // Only check relevent entries
+      if ( x1 < (data.len() + 2) ){
+        x-counter = x-counter + 1
+        ticks.push( (x-counter, x1 ))
+      }
+
+      if ( x2 < (data.len() + 2) ){
+        x-counter = x-counter + 1
+        ticks.push( (x-counter, x2 ))
+      }
+    }
+
+    x-ticks = ticks
+  }
+
+  let y-unit = y-unit
+  if y-unit == auto {
+    y-unit = []
+  }
+
+  let x = axes.axis(min: 0,
+                    max: data.len() + 2,
+                    label: x-label,
+                    ticks: (list: x-ticks,
+                            grid: none, step: none,
+                            minor-step: none,
+                    ))
+  let y = axes.axis(min: min-value, max: max-value,
+                    label: y-label,
+                    ticks: (grid: true, step: y-tick-step,
+                            minor-step: none,
+                            unit: y-unit, decimals: 1,
+                            list: y-ticks))
+
+  if mode == "horizontal" {
+    (x, y) = (y, x)
+  }
+
+  let vertical-draw-dendrogram(data, ..style) = {
+    let data-mut = data // Mutable
+    let line-style = line-style;
+    if type(line-style) != function { line-style = ((i) => line-style) }
+
+    let x-counter = 0
+    let x-array = (false,) * (data.len() + 1)
+
+    for (idx, entry) in data.enumerate() {
+      let height = entry.at(height-key)
+
+      let x1 = entry.at(x1-key)
+      let x2 = entry.at(x2-key)
+
+      let y1 = 0
+      let y2 = 0
+
+      if (x1 > (data.len() + 1)){
+          let child = data-mut.at(x1 - 2)
+          x1 = child.at(x1-key)
+          y1 = child.at(height-key)
+      } else {
+        // x2 is a ground-level leave
+        // Does it have an assigned x?
+        let possible-id = x-array.at(x1, default: false)
+        if ( possible-id == false ){
+          x-counter = x-counter + 1
+          x-array.insert(x1, x-counter)
+          x1 = x-counter
+        } else {
+          x1 = possible-id
+        }
+      }
+
+      if (x2 > (data.len()) + 1){
+          let child = data-mut.at(x2 - 2)
+          x2 = child.at(x1-key)
+          y2 = child.at(height-key)
+      } else {
+        // x2 is a ground-level leave
+        // Does it have an assigned x?
+        let possible-id = x-array.at(x2, default: false)
+        if ( possible-id == false ){
+          x-counter = x-counter + 1
+          x-array.insert(x2, x-counter)
+          x2 = x-counter
+        } else {
+          x2 = possible-id
+        }
+      }
+
+      line((x1, y1), (x1, height), (x2, height), (x2, y2),
+        ..style, ..line-style(idx))
+
+      data-mut.push((
+        (x1 + x2) / 2,
+        (x1 + x2) / 2,
+        height
+      ))
+    }
+  }
+
+  let horizontal-draw-dendrogram(data, ..style) = {
```

Comment:
is the only difference between this function and the vertical version the params to the `line` function? Could they be merged somehow?

Reply:
Done, moved the logic that figures out (x,y) coordinates to its own function.